test/version_solver_test.dart

```diff
@@ -1300,6 +1301,241 @@ void downgrade() {
+    await expectResolves(result: {'foo': '1.0.0', 'bar': '1.0.0'});
+  });
+
+  test("doesn't select a version with an incompatible feature", () async {
+    await servePackages((builder) {
+      builder.serve('foo', '1.0.0', pubspec: {
+        "features": {
+          "feature": {
+            "dependencies": {'bar': '1.0.0'}
+          }
+        }
+      });
+      builder.serve('foo', '1.1.0', pubspec: {
+        "features": {
+          "feature": {
+            "dependencies": {'bar': '2.0.0'}
+          }
+        }
+      });
+      builder.serve('bar', '1.0.0');
+      builder.serve('bar', '2.0.0');
+    });
+
+    await d.appDir({
+      'foo': {
+        'version': '^1.0.0',
+        'features': {'feature': true}
+      },
+      'bar': '1.0.0'
+    }).create();
+    await expectResolves(result: {'foo': '1.0.0', 'bar': '1.0.0'});
+  });
+
+  test("doesn't select a version with a transitively incompatible feature",
+      () async {
+    await servePackages((builder) {
+      builder.serve('foo', '1.0.0', pubspec: {
+        "features": {
+          "feature": {
+            "dependencies": {'bar': '1.0.0'}
+          }
+        }
+      });
+      builder.serve('foo', '1.1.0', pubspec: {
+        "features": {
+          "feature": {
+            "dependencies": {
+              'bar': {'version': '1.0.0', 'feature': 'bar_feature'}
+            }
+          }
+        }
+      });
+
+      builder.serve('bar', '1.0.0', pubspec: {
+        "features": {
+          "feature": {
+            "dependencies": {'baz': '1.0.0'}
+          }
+        }
+      });
+
+      builder.serve('baz', '1.0.0');
+      builder.serve('baz', '2.0.0');
+    });
+
+    await d.appDir({
+      'foo': {
+        'version': '^1.0.0',
+        'features': {'feature': true}
+      },
+      'baz': '2.0.0'
+    }).create();
+    await expectResolves(
+        result: {'foo': '1.0.0', 'bar': '1.0.0', 'baz': '2.0.0'});
+  });
+
+  test("backtracks if a feature's dependencies are transitively incompatible",
+      () async {
+    await servePackages((builder) {
+      builder.serve('foo', '1.0.0', pubspec: {
+        "features": {
+          "feature": {
+            "dependencies": {'bar': '1.0.0'}
+          }
+        }
+      });
+      builder.serve('foo', '1.1.0', pubspec: {
+        "features": {
+          "feature": {
+            "dependencies": {'bar': '2.0.0'}
+          }
+        }
+      });
+
+      builder.serve('bar', '1.0.0', deps: {'baz': '1.0.0'});
+      builder.serve('bar', '2.0.0', deps: {'baz': '2.0.0'});
+
+      builder.serve('baz', '1.0.0');
+      builder.serve('baz', '2.0.0');
+    });
+
+    await d.appDir({
+      'foo': {
+        'version': '^1.0.0',
+        'features': {'feature': true}
+      },
+      'baz': '1.0.0'
+    }).create();
+    await expectResolves(
+        result: {'foo': '1.0.0', 'bar': '1.0.0', 'baz': '1.0.0'}, tries: 2);
+  });
+
+  test("the root package's features are disabled by default", () async {
```

Comment:
Currently, only if a dependency enables it. We can talk about more explicit ways to enable them, but I think that's out of scope for this CL.